Archive/Imo/Imo1985Q2.lean

```diff
@@ -0,0 +1,99 @@
+/-
+Copyright (c) 2025 Jeremy Tan. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jeremy Tan
+-/
+import Mathlib.Algebra.Order.Ring.Canonical
+import Mathlib.Algebra.Order.Ring.Star
+import Mathlib.Data.Nat.ModEq
+
+/-!
+# IMO 1985 Q2
+
+Each of the numbers in the set $N=\{1, 2, 3, \dots, n-1\}$, where $n ≥ 3$, is colored with one of
+two colors, say red or black, so that:
+
+1. $i$ and $n-i$ always receive the same color, and
+2. for some $j ∈ N$ relatively prime to $n$, $i$ and $|j-i|$ receive the same color
+for all $i ∈ N, i ≠ j$.
+
+Prove that all numbers in $N$ must receive the same color.
+
+# Solution
+
+Let $a \sim b$ denote that $a$ and $b$ have the same color.
+Because $j$ is coprime to $n$, every number in $N$ is of the form $kj\bmod n$ for a unique
+$1 ≤ k < n$, so it suffices to show that $kj\bmod n \sim (k-1)j\bmod n$ for $1 < k < n$.
+In this range of $k$, $kj\bmod n ≠ j$, so
+
+* if $kj\bmod n > j$, $kj\bmod n \sim kj\bmod n - j = (k-1)j\bmod n$ using rule 2;
+* if $kj\bmod n < j$, $kj\bmod n \sim j - kj\bmod n \sim n - j + kj\bmod n = (k-1)j\bmod n$
+using rule 2 then rule 1.
+-/
+
+namespace Imo1985Q2
+
+open Nat
+
+/-- The conditions on the problem's coloring `C`.
+Although its domain is all of `ℕ`, we only care about its values in `Set.Ico 1 n`. -/
+def Condition (n j : ℕ) (C : ℕ → Bool) : Prop :=
+  (∀ i ∈ Set.Ico 1 n, C i = C (n - i)) ∧ (∀ i ∈ Set.Ico 1 n, i ≠ j → C i = C (j - i : ℤ).natAbs)
+
+lemma Int.natAbs_sub_nat_of_lt {a b : ℕ} (h : b ≤ a) : (a - b : ℤ).natAbs = a - b := by
+  omega
+
+lemma Int.natAbs_sub_nat_of_gt {a b : ℕ} (h : a ≤ b) : (a - b : ℤ).natAbs = b - a := by
+  omega
+
+lemma Nat.mod_sub_comm {a b n : ℕ} (h : b ≤ a % n) : a % n - b = (a - b) % n := by
+  rcases n.eq_zero_or_pos with rfl | hn; · simp only [mod_zero]
+  nth_rw 2 [← div_add_mod a n]; rw [Nat.add_sub_assoc h, mul_add_mod]
+  exact (mod_eq_of_lt <| (sub_le ..).trans_lt (mod_lt a hn)).symm
+
+/-- For `1 ≤ k < n`, `k * j % n` has the same color as `j`. -/
+lemma C_mul_mod {n j : ℕ} (hn : 3 ≤ n) (hj : j ∈ Set.Ico 1 n) (cpj : Coprime n j)
+    {C : ℕ → Bool} (hC : Condition n j C) {k : ℕ} (hk : k ∈ Set.Ico 1 n) :
+    C (k * j % n) = C j := by
+  rcases hk.1.eq_or_lt with rfl | (hk₁ : 1 + 1 ≤ k)
+  · rw [one_mul, mod_eq_of_lt hj.2]
+  · have nej : k * j % n ≠ j := by
+      by_contra! h; nth_rw 2 [← mod_eq_of_lt hj.2, ← one_mul j] at h
+      replace h : k % n = 1 % n := ModEq.cancel_right_of_coprime cpj h
+      rw [mod_eq_of_lt hk.2, mod_eq_of_lt (by omega)] at h
+      omega
+    have b₁ : k * j % n ∈ Set.Ico 1 n := by
+      refine ⟨?_, mod_lt _ (by omega)⟩
+      by_contra! h; rw [lt_one_iff, ← dvd_iff_mod_eq_zero] at h
+      have ek := eq_zero_of_dvd_of_lt (cpj.dvd_of_dvd_mul_right h) hk.2
+      omega
+    have hk₂ : k - 1 ∈ Set.Ico 1 n := ⟨le_sub_of_add_le hk₁, (sub_le ..).trans_lt hk.2⟩
+    rw [← C_mul_mod hn hj cpj hC hk₂, hC.2 _ b₁ nej]
```

Comment:
This self-referential `C_mul_mod` caught me by surprise. Could you either
 1. add a comment explaining which argument is getting smaller to make this recursive call work (I think it's `k`), or
 2. restructure the proof so that it explicitly uses the `induction` tactic?

Reply:
Both should be done.

Reply:
I tried writing a version that uses `induction` instead: https://github.com/dwrensha/compfiles/blob/59b8b76e579aa0a313c4f3c52051df9b7b273eff/Compfiles/Imo1985P2.lean#L63-L92
I'll let you decide whether you think that's any better.